### ISSUE TITLE
Generic classes cause massive slowdown or hang

### ISSUE THREAD

**Issue by RobertCraigie** [python/mypy#11484]
Hi, this [commit](https://github.com/RobertCraigie/prisma-client-py/pull/102/commits/fcc6e3520594f24a4d87ed3cf281b7b2704f959f), causes mypy to go from ~7 minutes to ~30 minutes to typecheck the whole codebase.

It should be noted that the code in the mentioned commit is used to auto-generate 5 classes, each with 9 methods that use the generic typevar. These methods don't do anything special but the arguments are heavily nested TypedDicts. The classes look like this:

```py
from pydantic import BaseModel

BaseModelT = TypeVar('BaseModelT', bound=BaseModel)

class UserActions(Generic[BaseModelT]):
    async def create(
        self,
        data: 'UserCreateInput',
    ) -> BaseModelT:
        ...
    ...
```

If it is still relevant, this https://github.com/python/mypy/issues/3796#issuecomment-320603042 might explain some of the reasons behind this performance loss.

I'm sorry I don't have a smaller example, I don't have enough time to dig any deeper into this issue.

**Comment by RobertCraigie** [comment]
I actually cannot run this on my machine, left it running for ~2 hours with no result.

**Comment by hauntsaninja** [comment]
Is there a standalone Python file that is slow to type check? I also see you have multiple mypy plugins. Does disabling them help?

**Comment by RobertCraigie** [comment]
@hauntsaninja I think the worst offender is the code generated from https://github.com/RobertCraigie/prisma-client-py/blob/main/src/prisma/generator/templates/types.py.jinja

On mobile right now but will share an example of the generated file when possible.

The auto-generated file is basically a lot of nested and pseudo-recursive TypedDicts, I think the last time I checked the size of the mypy cache for this file was around 400MB, that could be the issue?

**Comment by RobertCraigie** [comment]
generated `types.py` file: https://mystb.in/EvilPlanesPain.python

**Comment by RobertCraigie** [comment]
Mypy was slightly faster without the plugins, but still taking 27 minutes

https://github.com/RobertCraigie/prisma-client-py/runs/4273696907?check_suite_focus=true

**Comment by RobertCraigie** [comment]
The stand alone `types.py` file does not seem to be the root cause of this issue, type checking took around a minute:

https://github.com/RobertCraigie/prisma-client-py/runs/4280666648?check_suite_focus=true

**Comment by Fulguritude** [comment]
I'm in a very similar situation, and can confirm that generics and nested dicts/type aliases do not mesh well currently. I recently started to add a bunch of genericity to refactor some of our code. I've tried running mypy for more than 15 minutes, and mypy still hadn't returned its errors; so I was thinking this was possibly an issue on my end - until I came across this thread. Clearly it's taking an ungodly long time, compared to the usual minute or so that it takes to check our "non-generic" codebase.
